src/workerd/io/worker.c++

```diff
@@ -2953,89 +2966,77 @@ void Worker::Actor::Impl::HooksImpl::updateAlarmInMemory(kj::Maybe<kj::Date> new
+  return nullptr;
+}
 
-        // now we can replace the queued alarm with a new one. we exclusiveJoin with the paf promise
-        // to allow for future overwrites
-        return r->queuedAlarm
-          .emplace(makeQueuedAlarm(r->alarm.addBranch().ignoreResult()))
-          .alarm.addBranch();
-      }
-    } else {
-      // there's not a queued alarm already, so we're safe to just go ahead and set it.
-      return r->queuedAlarm
-        .emplace(makeQueuedAlarm(r->alarm.addBranch().ignoreResult()))
-        .alarm.addBranch();
+auto Worker::Actor::scheduleAlarm(kj::Date scheduledTime) -> kj::Promise<ScheduleAlarmResult> {
+  KJ_IF_MAYBE(runningAlarm, impl->maybeRunningAlarm) {
+    if (runningAlarm->scheduledTime == scheduledTime) {
+      // The running alarm has the same time, we can just wait for it.
+      auto result = co_await runningAlarm->resultPromise.addBranch();
+      co_return result;
     }
-  } else {
-    auto [prom, fulfiller] = kj::newPromiseAndFulfiller<WorkerInterface::AlarmResult>();
-    auto& fulfillerRef = *fulfiller;
-    auto& running = impl->runningAlarm.emplace(Impl::RunningAlarm {
-      Impl::Alarm {
-        runAlarmImpl(fulfillerRef),
-        prom.fork(),
-        kj::mv(fulfiller),
-        scheduledTime
-      }
-    });
-    return running.alarm.addBranch();
   }
+
+  KJ_IF_MAYBE(scheduledAlarm, impl->maybeScheduledAlarm) {
+      // We had a previously scheduled alarm, let's cancel it.
+      scheduledAlarm->resultFulfiller.cancel();
+      impl->maybeScheduledAlarm = nullptr;
+  }
+
+  KJ_IASSERT(impl->maybeScheduledAlarm == nullptr);
+  auto& scheduledAlarm = impl->maybeScheduledAlarm.emplace(
+      scheduledTime, kj::newPromiseAndFulfiller<AlarmResult>());
+
+  auto whenCanceled = scheduledAlarm.resultPromise.addBranch().then(
+      [](AlarmResult result) -> ScheduleAlarmResult {
+    // We've been cancelled, so return that result. Note that we cannot be resolved any other
+    // way until we return an AlarmFulfiller below.
+    return result;
+  });
+
+  // Let's wait for any running alarm to cleanup before we even delay.
+  auto whenReady = impl->runningAlarmTask.addBranch().then([this, scheduledTime](){
+    // Date.now() < scheduledTime when the alarm comes in, since we subtract elapsed CPU time from
+    // the time of last I/O in the implementation of Date.now(). This difference could be used to
+    // implement a spectre timer, so we have to wait a little longer until
+    // `Date.now() == scheduledTime`. Note that this also means that we could invoke ahead of its
+    // `scheduledTime` and we'll delay until appropriate, this may be useful in cases of clock skew.
+    return KJ_ASSERT_NONNULL(impl->ioContext)->atTime(scheduledTime);
+  });
+
+  co_return co_await whenReady.then([this]() -> ScheduleAlarmResult {
+    // It's time to run! Let's tear apart the scheduled alarm and make a running alarm.
+    auto scheduledAlarm = KJ_ASSERT_NONNULL(kj::mv(impl->maybeScheduledAlarm));
```

Comment:
So we don't need to verify that `impl->maybeScheduledAlarm` still contains the expected `scheduledTime`? Because `whenCanceled` would definitely be resolved first (before `whenReady`) if `scheduledAlarm` had been removed/changed?

That looks like it's probably true, I just haven't convinced myself it -- so mostly I'm asking for reassurance that you have convinced yourself.

Reply:
Yep, you have it right, but that deserves a comment!